pkg/controllers/workgenerator/controller.go

```diff
@@ -0,0 +1,329 @@
+	activeWork := make(map[string]bool, len(resourceSnapshots))
+	for _, snapshot := range resourceSnapshots {
+		workName := getWorkNameFromSnapshotName(snapshot.Name)
+		activeWork[workName] = true
+		// TODO(RZ): issue those requests in parallel to speed up the process
+		err = r.upsertWork(ctx, works[workName], snapshot, resourceBinding)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	//  delete the works that are not associated with any resource snapshot
+	for _, work := range works {
+		if activeWork[work.Name] {
+			continue
+		}
+		klog.V(2).InfoS("Delete the work that is not associated with any resource snapshot", "work", klog.KObj(work))
+		if err := r.Client.Delete(ctx, work); err != nil {
+			klog.ErrorS(err, "failed to delete the no longer needed work", "work", klog.KObj(work))
+			if !errors.IsNotFound(err) {
+				return ctrl.Result{}, controller.NewAPIServerError(false, err)
+			}
+		}
+	}
+
+	klog.V(2).InfoS("Successfully synced all the work associated with the resourceBinding", "resourceBinding", klog.KObj(resourceBinding))
+	return ctrl.Result{}, nil
+}
+
+// fetchAllResourceSnapshots gathers all the resource snapshots for the resource binding.
+func (r *Reconciler) fetchAllResourceSnapshots(ctx context.Context, resourceBinding *fleetv1beta1.ClusterResourceBinding) (map[string]*fleetv1beta1.ClusterResourceSnapshot, error) {
+	resourceSnapshots := make(map[string]*fleetv1beta1.ClusterResourceSnapshot, 1)
+	resourceSnapshot := fleetv1beta1.ClusterResourceSnapshot{}
+	if err := r.Client.Get(ctx, client.ObjectKey{Name: resourceBinding.Spec.ResourceSnapshotName}, &resourceSnapshot); err != nil {
+		klog.ErrorS(err, "failed to get the resource snapshot from resource resourceBinding",
+			"resource resourceBinding", klog.KObj(resourceBinding), "resource snapshot name", resourceBinding.Spec.ResourceSnapshotName)
+		return nil, controller.NewAPIServerError(true, err)
+	}
+	resourceSnapshots[resourceSnapshot.Name] = &resourceSnapshot
+
+	// check if there are more snapshot in the same index group
+	countAnnotation := resourceSnapshot.Annotations[fleetv1beta1.NumberOfResourceSnapshotsAnnotation]
+	snapshotCount, err := strconv.Atoi(countAnnotation)
+	if err != nil {
+		return nil, controller.NewUnexpectedBehaviorError(err)
+	}
+	if snapshotCount > 1 {
+		index, exist := resourceSnapshot.Labels[fleetv1beta1.ResourceIndexLabel]
+		if !exist {
+			unexpectedErr := fmt.Errorf("resource snapshot %s has no index label", resourceSnapshot.Name)
+			return nil, controller.NewUnexpectedBehaviorError(unexpectedErr)
+		}
+		resourceIndexLabelMatcher := client.MatchingLabels{
+			fleetv1beta1.ResourceIndexLabel: index,
+		}
+		resourceSnapshotList := &fleetv1beta1.ClusterResourceSnapshotList{}
+		if err := r.Client.List(ctx, resourceSnapshotList, resourceIndexLabelMatcher); err != nil {
+			klog.ErrorS(err, "failed to list all the resource snapshot associated with the resourceBinding", "resourceBinding", klog.KObj(resourceBinding))
```

Comment:
```suggestion
			klog.ErrorS(err, "Failed to list all the resource snapshot associated with the resourceBinding", "resourceBinding", klog.KObj(resourceBinding))
```

Reply:
I thought error should not start with capital letter

Reply:
oh, this rule does not apply to the log strings.

`so that log.Printf("Reading %s: %v", filename, err) formats without a spurious capital letter mid-message. This does not apply to logging, which is implicitly line-oriented and not combined inside other messages.`

https://github.com/golang/go/wiki/CodeReviewComments#error-strings